### ISSUE TITLE
BUG: multivariate_normal returns a pdf for values outside its support when covariance is singular

### ISSUE THREAD

**Issue by drpeteb** [scipy/scipy#5286]
When a multivariate normal distribution has a singular covariance matrix, its support (i.e. the set of possible values the random variable can take) is restricted to manifold. In this case, the density with respect to the usual Lebesgue measure is not defined. There IS a density with respect to the Hausdorff measure on the support, and this is exactly what stats.multivariate_normal returns. The problem is that it returns a pdf for any input. Specifically, it returns the density of the projection of the input onto the support. I don't think this behaviour is correct. To demonstrate:

```python
>>> import numpy as np
>>> from scipy.stats import multivariate_normal
>>> mean = np.zeros(2)
>>> cov = np.ones((2,2))
>>> rv = multivariate_normal(mean=mean, cov=cov, allow_singular=True)
>>> x = np.array([[0.0, 0.0], [1.0, -1.0], [2.0, -2.0]])
>>> x
array([[ 0.,  0.],
       [ 1., -1.],
       [ 2., -2.]])
>>> rv.pdf(x)
array([ 0.28209479,  0.28209479,  0.28209479])
```

For this distribution, the random variable must lie on the line `y = x`. `[0,0]` is within the support, so that's all well and good. The other two points are not. Technically, the density is simply undefined at these points. For consistency with other scipy.stats distributions, the output should probably be `0.0`.

**Comment by drpeteb** [comment]
I've never contributed to scipy before. I've patched this. Do you want it? Do I just submit a pull request? Should I add tests?

https://github.com/scipy/scipy/compare/master...drpeteb:bug-mvn-singular-covariance

**Comment by argriffing** [comment]
This is kind of a tricky issue and I'm not sure how it should have been handled.  I think I've understood what you've said about densities and manifolds and projections, and honestly it's kind of reassuring to hear that the current code at least behaves in an internally consistent (even if poorly chosen) way.

> Technically, the density is simply undefined at these points. 

Just to make sure I understand your proposed patch, the idea is that if a point is farther than some distance `eps` from the subspace then with your patch the pdf will return 0 (instead of the current behavior of using the projection onto the distant subspace), and if the point is closer than `eps` to the subspace then the patch will not change the current behavior?

**Comment by drpeteb** [comment]
Yes. That's exactly it. `eps` is the same value as is used to determine which eigenvalues are `0`. Looking at it again, there's a simpler method for getting the projection matrix with no additional matrix multiplications.

A simpler way to describe the issue is that it currently assigns a probability density to points that could never be generated by the `rvs()` method. A 1D equivalent would be like a gamma distribution giving a density to negative values.

**Comment by WarrenWeckesser** [comment]
A reasonable behavior when the covariance matrix is singular is to simply raise a `ValueError` (or some other appropriate error).  For example, consider the univariate case, where a singular covariance matrix corresonds to a variance of 0.  This case can be thought of as the limit of the standard distributions with the variance approaching 0.  As the variance goes to 0, the PDF at 0 goes to infinity.  The limiting case of the PDF would be a delta "function" at the origin.  I don't see how it could be considered correct to return 1 at 0 when the variance is 0.

**Comment by argriffing** [comment]
@WarrenWeckesser Yes, `allow_singular` is `False` by default, so it raises a LinAlgError on singular matrices. If someone wants to do exotic things involving Hausdorff measures they have to give the keyword argument explicitly.

@drpeteb I think I like the idea.  Technically it would break backwards compatibility; if someone was counting on the function to do the projection implicitly then this patch breaks their code.  Deprecations may be required.

> Do I just submit a pull request? Should I add tests?

Yes, that would be great!

**Comment by WarrenWeckesser** [comment]
@argriffing Whoops--sorry, I read this issue without going back to look at the code (and without noticing the `allow_singular` argument in the example).  The behavior being discussed for the case `allow_singular=True` seems a bit dicey, numerically speaking, given normal floating point imprecision.   A tiny change in a near-singular covariance matrix that makes the matrix closer to singular can result in PDF values changing from huge to tiny.  I don't really see the use-cases for this.  However, as you point out, it isn't the default behavior, so it won't affect most users.

**Comment by argriffing** [comment]
pinging `multivariate_normal` contributor @jvkersch ...

**Comment by drpeteb** [comment]
@WarrenWeckesser I agree - dicey. I don't like the fact that it gets to decide what the rank is for exactly that reason.

**Comment by josef-pkt** [comment]
@drpeteb  What is the usecase for the degenerate pdf?

I thought the initial proposal was quite reasonable, but I start to see the problems in the interpretation that Warren is pointing out.
For practical purposes I guess that the initial solution is still more likely the expected and useful solution.

I recently saw a similar case where the behavior at the threshold that Warren points out causes weird behavior
http://stackoverflow.com/questions/32599159/robustness-issue-of-statsmodel-linear-regression-ols-python  in that case it is between inv and pinv at the rcond threshold when the reduced rank treatment kicks in.

Also, we have several cases with degenerate multivariate normal distribution (reduced rank cov_params) in statsmodels. I started to worry about the behavior similar to this issue where we would have to look at the appropriate subspace. But because the behavior that we actually use (like the non-degenerate distribution of linear combinations), I didn't need to make up my mind about the desired behavior.

**Comment by josef-pkt** [comment]
I think the solution is theoretically clean if we define it as the distribution on the subspace, which is the main application in statsmodels of similar degenerate models, e.g. models with linear restrictions, or zero restrictions which have zero variance for some dimensions or subspaces.

The only problem is the numerical issue in the threshold. But for this I think "practicality beats purity".  An almost degenerate distribution has very bad numerical properties and users have to make up their mind whether they want near-singular (eps=0) or singular behavior (eps>0). Same problem is in pinv, linalg.matrix_rank and so on. But the defaults work in 99.99 % (?) of the cases.

**Comment by WarrenWeckesser** [comment]
Josef, I think we're on the same wavelength.  I was writing the following as your comment appeared.

Here's some clarification of my comment about not seeing the use-cases for `allow_singular=True`.

What I *don't* see a use-case for is when a user calls the pdf (or any other) method with `allow_singular=True` and with a covariance matrix that might or might not be singular.  If the method does the right thing when the covariance matrix nonsingular or near-singular (but not numerically singular), how could the singular case (which drastically changes the PDF) *also* be what the user wants?

"In that case," one might say, "the user should be using the default, `allow_singular=False`."  But then who uses `allow_singular=True`?  It would be used when *the user already knows the covariance matrix is singular* (up to normal floating point precision, of course).  In other words, it is really a separate mathematical and numerical problem: defining a probability distribution on a subspace.  In this case, the *domain* of the distribution is the subspace, and, as @drpeteb said above, the distribution is not even defined off of the subspace.  (E.g. it is the integral of the pdf over the *subspace* that gives 1, rather than the integral over the entire space in which the subsapce lies.)

Given that, I think it would make more sense to define a *separate class* for distributions on subspaces.  (I'd also remove the `allow_singular` option from `multivariate_normal`, but it is already out there, so removing it might be more trouble than keeping it.)

**Comment by drpeteb** [comment]
We're all on the same wavelength. I would always recommend handling a singular covariance matrix by keeping it factorised, so that the rank is known explicitly at all times. Then if you want the PDF you map the subspace into its lower dimension first. That would mean specifying the covariance with two parameters: one defines the subspace, the other is the "within-subspace" covariance. Definitely a job for a new class.

There's two issues here. First, the current behaviour is incorrect, or at least highly dubious, which my patch is designed to fix. Second, there's the wider issue of degenerate distributions. Should we quick-fix the first? The second might be a substantially larger job.

My current patch breaks the test_degenerate_distributions test, for obvious reasons.

**Comment by josef-pkt** [comment]
I still think that the degenerate option and distribution are useful. But it would be helpful to have methods to get to a full rank subspace distribution.

In the statsmodels usecase, we give the degenerate cov_params to the user and things can be directly calculated with it. In the internal estimation, we use the restricted (subspace) version.

Simple example, we have 10 variables and set some of them to zero, so those have zero variance, fixed at a specific value. 
(For some tasks we also need both an unrestricted full rank covariance to be compared with the covariance when it's projected on the subspace, i.e. comparing a unrestricted with a restricted model.)

So, I think having the singular option is still useful, but we never needed so far to evaluate something like the pdf of the degenerate case on points that are not in the support/subspace (given that we control the usage and in contrast to scipy don't directly expose it as a function or class). 

Conclusion, I'm in favor of the proposed change.

**Comment by josef-pkt** [comment]
Two more thoughts

- getting more support for moving between degenerate and full rank distributions would be useful, maybe in a new class as proposed. Especially I guess that would allow to have a proper definition for all methods, if there are other shaky ones.
- As Warren pointed out with the comparison to the univariate case with variance equal to zero: In the multivariate case with degenerate distribution we are still within the same class of multivariate distributions but defined on a subspace. However, if the restriction is so that the degenerate distribution collapses to a single point, then we might run into the same problems as in the univariate case. So maybe the one point case still needs to be excluded. (I haven't checked what the current code does.)

**Comment by drpeteb** [comment]
As it is, with `cov=0`, the pdf returned is `1.0` everywhere, which is consistent at least. With the patch it's `1.0` when `x=mean` and `0.0` everywhere else.

**Comment by WarrenWeckesser** [comment]
@josef-pkt: The origin is a 0-dimensional subspace of the line, so the univariate example is still within the same class of problems as the multivariate cases.

**Comment by drpeteb** [comment]
One possible option for making the degenerate behaviour more robust would be to allow the user to specify the expected rank / dimension of the subspace. Then if the numerical rank disagrees, raise an error. If it's not specified, detect it automatically (i.e. what it currently does).

**Comment by argriffing** [comment]
> That would mean specifying the covariance with two parameters: one defines the subspace, the other is the "within-subspace" covariance. Definitely a job for a new class.

Another possible way to do this for a hypothetical new class could be to have an argument `X` to replace `cov`, such that `cov = X.T.dot(X)`. I think this could let `X` simultaneously specify the subspace (including its intended rank) and the within-subspace covariance, while also possibly being more numerically stable in the common use-case where the covariance matrix would have been constructed in this manner anyway.

**Comment by mdhaber** [comment]
This issue is being used to track gh-9257, gh-13541. If those are really distinct, they should be reopened, but it would be great if we could address these all in one go.

**Comment by mdhaber** [comment]
[Those issues](https://github.com/scipy/scipy/issues/5286#issuecomment-1068773655) were somewhat distinct, but both received responses that explained what was happening and why it was reasonable. There were no clear bugs or steps for improvement, so I think it was safe to close them.

While working on gh-15509, I noticed the behavior reported in this issue.

**Comment by mdhaber** [comment]
This sequence of plots illustrates the issue.  `stdx` and `stdy` are the nonzero elements of a diagonal covariance matrix.

<img width="464" alt="image" src="https://user-images.githubusercontent.com/6570539/163094428-8351d646-895b-4349-9f60-2848b9d62c2c.png">
<img width="460" alt="image" src="https://user-images.githubusercontent.com/6570539/163094518-0bc9ba4e-6968-4179-87ec-fa6e4096943b.png">
<img width="453" alt="image" src="https://user-images.githubusercontent.com/6570539/163094576-8037153e-bb76-45e7-ab92-f5daf378e9e8.png">
<img width="456" alt="image" src="https://user-images.githubusercontent.com/6570539/163094822-b7484e30-c9b8-40f2-8b4d-65514382fe87.png">

**Comment by mdhaber** [comment]
A survey of multivariate normal with singular covariance matrix in other software.

Matlab doesn't allow it (see [here](https://www.mathworks.com/help/stats/mvnpdf.html))
<img width="937" alt="image" src="https://user-images.githubusercontent.com/6570539/163109367-87d7d7c6-1a47-4c40-bb78-778aed6b6d4c.png">

Mathematica doesn't allow it:
<img width="546" alt="image" src="https://user-images.githubusercontent.com/6570539/163109273-272314dc-e8fd-4af5-b7b9-552fc5995936.png">

@IgnacioHeredia based on [this](https://github.com/scipy/scipy/issues/15509#issue-1122145870), it sounded like you were saying that R's `mvnorm` was just wrong?

**Comment by IgnacioHeredia** [comment]
> I thought you might be interested in this. How were you expecting `multivariate_normal` PDF to behave in the singular case? 

Hi @mdhaber, I don't have a clear intuition on how it should work, because I was just only recently involved with this topic tangentially, nor have I a strong statistics background. Therefore, take my opinion with a grain of salt. 

The last plot you showed appears to be indeed quite counter-intuitive. I was waiting for some kind of Dirac-delta function appearing in the `y` axis (that is, a `NaN` flat plot in matplotlib). I guess this is consisting with what @argriffing was [pointing out](https://github.com/scipy/scipy/issues/5286#issuecomment-143033539) about Scipy projecting outside the subspace.

> How does the other software you used behave?

Python is my main software (I was toying with `R` when I posted the [issue](https://github.com/scipy/scipy/issues/15509#issue-1122145870), more below) so I cannot comment on other ones.

**Comment by IgnacioHeredia** [comment]
In the case of the `mvnorm` [package](https://www.rdocumentation.org/packages/mvtnorm/versions/1.1-3/topics/Mvnorm) in R, I was having the issue when computing the logpdf in huge arrays. I'm suspecting my issue was somewhat more connected to numerial precision (as [discussed above](https://github.com/scipy/scipy/issues/5286#issuecomment-143090548)).
To come back to the current issue, the output of `mvnorm` is (_more or less_) in line with what @drpeteb was [suggesting](https://github.com/scipy/scipy/issues/5286#issue-108197459) the correct behaviour should be:
```R
install.packages('mvtnorm')
library(mvtnorm)

mean <- c(0,0)
sigma <- matrix(c(1,1,1,1), ncol=2)

dmvnorm(c(0,0), mean=mean, sigma=sigma)   # inf
dmvnorm(c(1,-1), mean=mean, sigma=sigma)  # 0
dmvnorm(c(2,-2), mean=mean, sigma=sigma)  # 0
```
So I guess the behaviour (for values **inside** its support) is midway. It doesn't throw an error (as Matlab, or Mathematica) but neither gives a finite value corresponding to the Hausdorff measure on the subspace (as Scipy).
You can test it yourself in a [Colab with R](https://colab.research.google.com/#create=true&language=r).

All in all, based on my limited understanding, I think @drpeteb 's patch makes sense, along with the [possible option](https://github.com/scipy/scipy/issues/5286#issuecomment-143105451) to control numerical precision issues regarding the intended rank.

**Comment by mdhaber** [comment]
@IgnacioHeredia 
> To come back to the current issue, the output of `mvnorm` is (*more or less*) in line with what @drpeteb was [suggesting](https://github.com/scipy/scipy/issues/5286#issue-108197459) the correct behaviour should be:

Well, they agree that for values outside the support, the PDF is 0. But you write that `dmvnorm(c(0,0), mean=mean, sigma=sigma)` returns `inf` whereas the patch would have SciPy return a finite number - specifically, the PDF of the distribution defined on the lower-dimensional subspace spanned by the covariance matrix.

Does that still sound OK?
